lib/vmpooler/pool_manager.rb

```diff
@@ -683,36 +752,59 @@ def _check_pool(pool, provider)
-      $metrics.gauge("running.#{pool['name']}", $redis.scard("vmpooler__running__#{pool['name']}"))
-
-      if $redis.get("vmpooler__empty__#{pool['name']}")
-        $redis.del("vmpooler__empty__#{pool['name']}") unless ready.zero?
-      elsif ready.zero?
-        $redis.set("vmpooler__empty__#{pool['name']}", 'true')
-        $logger.log('s', "[!] [#{pool['name']}] is empty")
-      end
-
-      if total < pool['size']
-        (1..(pool['size'] - total)).each do |_i|
-          if $redis.get('vmpooler__tasks__clone').to_i < $config[:config]['task_limit'].to_i
-            begin
-              $redis.incr('vmpooler__tasks__clone')
-              pool_check_response[:cloned_vms] += 1
-              clone_vm(pool, provider)
-            rescue => err
-              $logger.log('s', "[!] [#{pool['name']}] clone failed during check_pool with an error: #{err}")
-              $redis.decr('vmpooler__tasks__clone')
-              raise
+      # Do not attempt to repopulate a pool while a template is updating
+      unless $redis.hget('vmpooler__config__updating', pool['name'])
+        ready = $redis.scard("vmpooler__ready__#{pool['name']}")
+        total = $redis.scard("vmpooler__pending__#{pool['name']}") + ready
+
+        $metrics.gauge("ready.#{pool['name']}", $redis.scard("vmpooler__ready__#{pool['name']}"))
+        $metrics.gauge("running.#{pool['name']}", $redis.scard("vmpooler__running__#{pool['name']}"))
+
+        if $redis.get("vmpooler__empty__#{pool['name']}")
+          $redis.del("vmpooler__empty__#{pool['name']}") unless ready.zero?
+        elsif ready.zero?
+          $redis.set("vmpooler__empty__#{pool['name']}", 'true')
+          $logger.log('s', "[!] [#{pool['name']}] is empty")
+        end
+
+        # Check to see if a pool size change has been made via the configuration API
+        # Since check_pool runs in a loop it does not
+        # otherwise identify this change when running
+        if $redis.hget('vmpooler__config__poolsize', pool['name'])
+          unless $redis.hget('vmpooler__config__poolsize', pool['name']).to_i == pool['size']
+            pool['size'] = Integer($redis.hget('vmpooler__config__poolsize', pool['name']))
+          end
+        end
+
+        if total < pool['size']
+          (1..(pool['size'] - total)).each do |_i|
+            if $redis.get('vmpooler__tasks__clone').to_i < $config[:config]['task_limit'].to_i
+              begin
+                $redis.incr('vmpooler__tasks__clone')
+                pool_check_response[:cloned_vms] += 1
+                clone_vm(pool, provider)
+              rescue => err
+                $logger.log('s', "[!] [#{pool['name']}] clone failed during check_pool with an error: #{err}")
+                $redis.decr('vmpooler__tasks__clone')
+                raise
+              end
             end
           end
         end
       end
 
+      # Remove VMs in excess of the configured pool size
+      remove_excess_vms(pool, provider, ready, total)
```

Comment:
The problem again is that we could hit this twice, and say each thread removes the extra 2 Vms over the total, we could be down 4 Vms. One way to reduce this exposure would be to add locks or to have the total and ready numbers checked for every single vm removal.

Reply:
It happens fast enough that I don't think this ends up being a concern in real world usage. We can try to cause a failure here and if we can make one happen we could try to implement some safeguards. With that said I have found that I have not been able to cause this to make a mistake purposefully.

Reply:
I have updated this to use a mutex object to synchronize these events so that only a single instance can happen across all threads. This should be more robust, and safe, than the redis based checking. I don't think that the methods used would have been likely to cause concern, but nonetheless this is more specific and should ensure the expected behavior.